File: app/src/main/java/com/wmods/wppenhacer/xposed/features/listeners/ConversationItemListener.java
package com.wmods.wppenhacer.xposed.features.listeners;

import android.app.Activity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.HeaderViewListAdapter;
import android.widget.ListAdapter;
import android.widget.ListView;

import androidx.annotation.NonNull;

import com.wmods.wppenhacer.xposed.core.Feature;
import com.wmods.wppenhacer.xposed.core.WppCore;
import com.wmods.wppenhacer.xposed.core.components.FMessageWpp;

import java.util.HashSet;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XSharedPreferences;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;

public class ConversationItemListener extends Feature {

    public static final HashSet<OnConversationItemListener> conversationListeners = new HashSet<>();
    private static ListAdapter mAdapter;
    private static XC_MethodHook.Unhook hooked;

    public ConversationItemListener(@NonNull ClassLoader loader, @NonNull XSharedPreferences preferences) {
        super(loader, preferences);
    }

    public static ListAdapter getAdapter() {
        return mAdapter;
    }

    @Override
    public void doHook() throws Throwable {
        XposedHelpers.findAndHookMethod(ListView.class, "setAdapter", ListAdapter.class, new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                Activity currentActivity = WppCore.getCurrentActivity();
                if (currentActivity == null || !currentActivity.getClass().getSimpleName().equals("Conversation")) {
                    return;
                }

                ListView listView = (ListView) param.thisObject;
                if (listView.getId() != android.R.id.list) {
                    return;
                }

                ListAdapter adapter = (ListAdapter) param.args[0];
                if (adapter instanceof HeaderViewListAdapter) {
                    adapter = ((HeaderViewListAdapter) adapter).getWrappedAdapter();
                }

                if (adapter == null) {
                    return;
                }

                mAdapter = adapter;

                for (OnConversationItemListener listener : conversationListeners) {
                    listener.onAttachAdapter(mAdapter);
                }

                if (hooked != null) {
                    hooked.unhook();
                }

                var method = mAdapter.getClass().getDeclaredMethod("getView", int.class, View.class, ViewGroup.class);
                hooked = XposedBridge.hookMethod(method, new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                        if (param.thisObject != mAdapter) return;

                        var position = (int) param.args[0];
                        var convertView = (View) param.args[1];
                        var viewGroup = (ViewGroup) param.getResult();

                        if (viewGroup == null) return;

                        Object fMessageObj = mAdapter.getItem(position);
                        if (fMessageObj == null) return;

                        var fMessage = new FMessageWpp(fMessageObj);

                        for (OnConversationItemListener listener : conversationListeners) {
                            try {
                                listener.onItemBind(fMessage, viewGroup, position, convertView);
                            } catch (Throwable e) {
                                logDebug(e);
                            }
                        }
                    }
                });
            }
        });
    }

    @NonNull
    @Override
    public String getPluginName() {
        return "Conversation Item Listener";
    }

    public abstract static class OnConversationItemListener {
        /**
         * Called when a message item is rendered in the conversation
         *
         * @param fMessage The message
         * @param view     The view associated with the item
         * @param position The position
         * @param convertView The view from the adapter
         * @throws Throwable Errors caught in the hook
         */
        public abstract void onItemBind(FMessageWpp fMessage, ViewGroup view, int position, View convertView) throws Throwable;

        public void onAttachAdapter(ListAdapter adapter) {

        }
    }
}

File: app/src/main/java/com/wmods/wppenhacer/xposed/features/listeners/ConversationItemListener.kt
package com.wmods.wppenhacer.xposed.features.listeners

import android.app.Activity
import android.os.Handler
import android.os.Looper
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.HeaderViewListAdapter
import android.widget.ListAdapter
import android.widget.ListView
import com.wmods.wppenhacer.xposed.core.Feature
import com.wmods.wppenhacer.xposed.core.WppCore
import com.wmods.wppenhacer.xposed.core.components.FMessageWpp
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XSharedPreferences
import de.robv.android.xposed.XposedBridge
import de.robv.android.xposed.XposedHelpers
import java.util.WeakHashMap

class ConversationItemListener(
    loader: ClassLoader,
    preferences: XSharedPreferences
) : Feature(loader, preferences) {

    data class BoundConversationItem(
        val messageId: String,
        val rowId: Long,
        val message: FMessageWpp
    )

    companion object {
        private const val FIELD_BOUND_MESSAGE_ID = "conversation_item_bound_message_id"

        @JvmField
        val conversationListeners = HashSet<OnConversationItemListener>()

        var adapter: ListAdapter? = null

        @JvmField
        val listItems = WeakHashMap<View, BoundConversationItem>()

        private var hooked: XC_MethodHook.Unhook? = null

        @JvmStatic
        fun notifyDataSetChanged() {
            Handler(Looper.getMainLooper()).post {
                (adapter as? BaseAdapter)?.notifyDataSetChanged()
            }
        }

        fun getBoundMessageId(view: View): String? {
            return XposedHelpers.getAdditionalInstanceField(view, FIELD_BOUND_MESSAGE_ID) as? String
        }

        fun isViewBoundToMessage(view: View, messageId: String): Boolean {
            return getBoundMessageId(view) == messageId
        }

        private fun bindViewToMessage(view: View, fMessage: FMessageWpp): BoundConversationItem {
            val boundItem = BoundConversationItem(
                messageId = fMessage.key.messageID,
                rowId = fMessage.rowId,
                message = fMessage
            )
            XposedHelpers.setAdditionalInstanceField(view, FIELD_BOUND_MESSAGE_ID, boundItem.messageId)
            listItems[view] = boundItem
            return boundItem
        }
    }

    @Throws(Throwable::class)
    override fun doHook() {
        WppCore.addListenerActivity { activity, type ->
            if (activity.javaClass.simpleName == "Conversation" && type == WppCore.ActivityChangeState.ChangeType.DESTROYED)
                hooked?.unhook()
        }

        XposedHelpers.findAndHookMethod(
            ListView::class.java,
            "setAdapter",
            ListAdapter::class.java,
            object : XC_MethodHook() {
                @Throws(Throwable::class)
                override fun beforeHookedMethod(param: MethodHookParam) {
                    val currentActivity = WppCore.getCurrentActivity()
                    if (currentActivity == null || currentActivity.javaClass.simpleName != "Conversation") {
                        return
                    }

                    val listView = param.thisObject as ListView
                    if (listView.id != android.R.id.list) {
                        return
                    }

                    var currentAdapter = param.args[0] as? ListAdapter
                    if (currentAdapter is HeaderViewListAdapter) {
                        currentAdapter = currentAdapter.wrappedAdapter
                    }

                    if (currentAdapter == null) {
                        return
                    }

                    adapter = currentAdapter

                    for (listener in conversationListeners) {
                        listener.onAttachAdapter(adapter)
                    }

                    hooked?.unhook()

                    val method = adapter!!.javaClass.getDeclaredMethod(
                        "getView",
                        Int::class.javaPrimitiveType,
                        View::class.java,
                        ViewGroup::class.java
                    )

                    hooked = XposedBridge.hookMethod(method, object : XC_MethodHook() {
                        @Throws(Throwable::class)
                        override fun afterHookedMethod(param: MethodHookParam) {
                            if (param.thisObject !== adapter) return

                            val position = param.args[0] as Int
                            val convertView = param.args[1] as? View
                            val viewGroup = param.result as? ViewGroup ?: return

                            val fMessageObj = adapter!!.getItem(position) ?: return

                            val fMessage = FMessageWpp(fMessageObj)

                            bindViewToMessage(viewGroup, fMessage)

                            for (listener in conversationListeners) {
                                try {
                                    listener.onItemBind(fMessage, viewGroup, position, convertView)
                                } catch (e: Throwable) {
                                    logDebug(e)
                                }
                            }
                        }
                    })
                }
            }
        )
    }

    override fun getPluginName(): String {
        return "Conversation Item Listener"
    }

    abstract class OnConversationItemListener {
        /**
         * Called when a message item is rendered in the conversation
         *
         * @param fMessage The message
         * @param view     The view associated with the item
         * @param position The position
         * @param convertView The view from the adapter
         * @throws Throwable Errors caught in the hook
         */
        @Throws(Throwable::class)
        abstract fun onItemBind(
            fMessage: FMessageWpp,
            view: ViewGroup,
            position: Int,
            convertView: View?
        )

        open fun onAttachAdapter(adapter: ListAdapter?) {
            // TODO
        }
    }
}
